### PR TITLE
Add grid sampling

### PR DESCRIPTION
**Pull request by pxl-th**
This PR adds [grid sampling](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html) for the 4D case.

This is needed for things like self-supervised depth estimation, where a neural network would learn a depth `D` for the current image and projection transformation `P` for the adjacent images and does "warping" of the images using `D` and `P`.
And this function does image sampling given projected coordinates (a.k.a. `grid`).

The implementation differs from the PyTorch's version in that it assumes `align_corners = True` and uses only bilinear interpolation. It was tested to give the same results as PyTorch's version.

[PR](https://github.com/FluxML/NNlibCUDA.jl/pull/31) with the CUDA kernels.

### CODE REVIEW

**Comment by pxl-th** [comment]
I've commented out failing test for the depthwise conv for now. Should I bring it back?

**Comment by CarloLucibello** [comment]
They were failing intermittently, mostly in the multi-threading environment. Since we have a few multithreaded methods, better have the multithreaded CI passing, so I'm ok with keeping those problematic tests out.  We should open an issue for those depthwiseconv failures if there is no one already. 

I'll try to review in the next few days, thanks for this contribution, at a first glance looks perfect already!

**Comment by ToucheSir** [comment]
@CarloLucibello RE CI failures, feel free to rename and re-purpose https://github.com/FluxML/NNlib.jl/issues/359.

File: src/sampling.jl [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+@inline ∇compute_source_index(coordinate, dim_size, ::Val{:zeros}) = ∇unnormalize(coordinate, dim_size)
+@inline function ∇compute_source_index(coordinate, dim_size, ::Val{:border})
+    source_coordinate, grad_in = ∇unnormalize(coordinate, dim_size)
+    source_coordinate, grad_clip = ∇clip_coordinate(source_coordinate, dim_size)
+    source_coordinate, grad_in * grad_clip
+end
+
+"""
+    grid_sample(input::AbstractArray{T, 4}, grid::AbstractArray{T, 4}; padding_mode = Val(:zeros))
```
Comment (CarloLucibello) [inline src/sampling.jl:29]:
```suggestion
    grid_sample(input::AbstractArray{T, 4}, grid::AbstractArray{T, 4}; padding_mode = :zeros)
```
Our API should not require users to wrap symbols in a `Val`

File: src/sampling.jl [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+(i.e. align corners is `true`).
+
+# Arguments
+
+- `input`: Input array in `(W_in, H_in, C, N)` shape.
+- `grid`: Input grid in `(2, W_out, H_out, N)` shape.
+    Where for each `(W_out, H_out, N)` grid contains `(x, y)`
+    coordinates that specify sampling locations normalized by the `input` shape.
+
+    Therefore, it should have values mostly in `[-1, 1]` range.
```
Comment (CarloLucibello) [inline src/sampling.jl:45]:
```suggestion
    Therefore, `x` and `y`  should have values mostly in `[-1, 1]` range.
```
Also, why "mostly?"

File: src/sampling.jl [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    For example, values `x = -1, y = -1` is the left-top pixel of `input`,
+    and values `x = 1, y = 1` is the right-bottom pixel of `input`.
```
Comment (CarloLucibello) [inline src/sampling.jl:47]:
```suggestion
    For example,  `(x = -1, y = -1)` is the left-top pixel of `input`,
    and `(x = 1, y = 1)` is the right-bottom pixel of `input`.
```

File: src/sampling.jl [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+    Out-of-bound values are handled accroding to the `padding_mode`.
```
Comment (CarloLucibello) [inline src/sampling.jl:49]:
```suggestion
    Out-of-bound values are handled according to the `padding_mode`.
```

File: src/sampling.jl [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        end
+        if in_bounds(iy_se, ix_se, iH, iW)
+            r += input[ix_se, iy_se, c, n] * se
+        end
+        output[w, h, c, n] = r
+    end
+end
+
+"""
+    ∇grid_sample(Δ::AbstractArray{T, 4}, input::AbstractArray{T, 4}, grid::AbstractArray{T, 4}; padding_mode = Val(:zeros)) where T
```
Comment (CarloLucibello) [inline src/sampling.jl:155]:
```suggestion
    ∇grid_sample(Δ::AbstractArray{T, 4}, input::AbstractArray{T, 4}, grid::AbstractArray{T, 4}; padding_mode = :zeros) where T
```

File: src/sampling.jl [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+end
+
+@inline function _safe_add!(dx, value, ix, iy, c, n)
+    @inbounds dx[ix, iy, c, n] += value
+end
+
+function rrule(::typeof(grid_sample), x, grid; padding_mode)
+    y = grid_sample(x, grid; padding_mode=padding_mode)
+    function grid_sample_pullback(Δ)
+        ∇x, ∇grid = @thunk(∇grid_sample(unthunk(Δ), x, grid; padding_mode=padding_mode))
```
Comment (CarloLucibello) [inline src/sampling.jl:249]:
```suggestion
        ∇x, ∇grid = ∇grid_sample(unthunk(Δ), x, grid; padding_mode=padding_mode)
```
Ideally, we want to separate the two gradient computations into 2 different thunks. Since the computation is shared here, wrapping into a thunk should be entirely useless

File: src/sampling.jl (hunk 2 above)
Comment (pxl-th) [reply src/sampling.jl]: >Also, why "mostly?"

It was added as a hint that the values can be outside, but probably makes no sense.

**Comment by maxfreu** [comment]
Hi! Just to leave this here, as I ported the tri/bi/linear upsampling code from pytorch: 1) Setting `align_corners = true` is not the best choice, as the gradients change with changing input image size. Pytorch's default therefore is `false`. But that doesn't change any results much. 2) The code you wrote reminds me a lot of what I wrote for upsampling. The kernels probably can't be shared, although they look quite similar. But the [utility functions](https://github.com/FluxML/NNlib.jl/blob/063992bc295568551395207f99e135dd1e802138/src/upsample.jl#L84-L98) can be shared I think - but that's maybe also for the future.

**Comment by gRox167** [comment]
Just to confirm, is there any plan to add 5D support for grid_sampling? 
This is really helpful for doing 3D image registration, which is widely use in neuro and medical imaging field. Networks in medical image registration utilize this function extensively to wrap the moving image to a fixed image, for example VoxelMorph.

**Comment by CarloLucibello** [comment]
No active plans that I know of, but contribution are welcome. Maybe @maxfreu could provide some guidance.

**Comment by maxfreu** [comment]
Hi @gRox167 ! As imaging science phd student this is the perfect task for you :D In principle it should be quite simple:

1. Fork NNlib
2. Come up with simple test cases for the forward and backward pass that you want to pass and write those first.
3. Add specializations for `AbstractArray{T,4}` to the current implementation where it is needed.
4. Copy the code for the 4D case, change the specialization to 5D.
5. Add the depth dimension where sizes are checked etc.
6. Change the sampling code in the kernel. You can use the pytorch code as a reference, which you can find [here](https://github.com/pytorch/pytorch/blob/f7000350905be5073892e0b23df681c0281be0f0/aten/src/ATen/native/cuda/GridSampler.cu#L156) for the GPU and [here](https://github.com/pytorch/pytorch/blob/f7000350905be5073892e0b23df681c0281be0f0/aten/src/ATen/native/GridSampler.cpp#L41) for the CPU.
7. Ping me or @pxl-th in case you hit bumps in the road :)
8. Enjoy how little code remains of the C++ hell.

It should be pretty straight forward and more of a writing than a thinking task. Optionally, you can simply use ChatGPT, show it the current julia and the pytorch code and tell it to expand it. I guess that should also work pretty well nowadays. The key to success in any case are good tests for the forward and backward pass.

**Comment by gRox167** [comment]
> Hi @gRox167 ! As imaging science phd student this is the perfect task for you :D In principle it should be quite simple:
> 
>     1. Fork NNlib
> 
>     2. Come up with simple test cases for the forward and backward pass that you want to pass and write those first.
> 
>     3. Add specializations for `AbstractArray{T,4}` to the current implementation where it is needed.
> 
>     4. Copy the code for the 4D case, change the specialization to 5D.
> 
>     5. Add the depth dimension where sizes are checked etc.
> 
>     6. Change the sampling code in the kernel. You can use the pytorch code as a reference, which you can find [here](https://github.com/pytorch/pytorch/blob/f7000350905be5073892e0b23df681c0281be0f0/aten/src/ATen/native/cuda/GridSampler.cu#L156) for the GPU and [here](https://github.com/pytorch/pytorch/blob/f7000350905be5073892e0b23df681c0281be0f0/aten/src/ATen/native/GridSampler.cpp#L41) for the CPU.
> 
>     7. Ping me or @pxl-th in case you hit bumps in the road :)
> 
>     8. Enjoy how little code remains of the C++ hell.
> 
> 
> It should be pretty straight forward and more of a writing than a thinking task. Optionally, you can simply use ChatGPT, show it the current julia and the pytorch code and tell it to expand it. I guess that should also work pretty well nowadays. The key to success in any case are good tests for the forward and backward pass.

Thanks for your guidance, I think I can take a try.

**Comment by gRox167** [comment]
Please refer to PR #627 .
